### PR TITLE
DMP-4061: InboundToUnstructured - Introduce Batch

### PR DESCRIPTION
**Pull request by Ben-Edwards-cgi**
### Change description ###
# Summary of Changes
A new parameter for specifying the batch size has been introduced to reduce the amount of data processed for each execution.
Updated task execution to asynchronously process each entry.

## Highlights

- **Batch Size Parameter**: The method `processInboundToUnstructured()` has been updated to accept an `int batchSize` parameter, allowing the system to process a specified number of records at once instead of processing them all in one go.

- **Configuration Changes**: The configuration file has been updated to remove the previous hardcoded limit for batch processing and set a default value for the batch size directly within the code.

- **Asynchronous Processing**: New utility functions have been introduced to handle asynchronous execution, which improves the efficiency of processing multiple records concurrently.

- **Database Migration**: A new SQL migration file has been added to ensure the database reflects the updated batch processing size for the automated task.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/darts/datamanagement/service/impl/InboundToUnstructuredProcessorImpl.java [hunk 1]
```diff
@@ -53,24 +50,47 @@ public class InboundToUnstructuredProcessorImpl implements InboundToUnstructured
...
-            log.debug("Processing Inbound to Unstructured record {} of {} with EOD {}", count++, inboundList.size(), inboundObject);
-            try {
-                singleElementProcessor.processSingleElement(inboundObject);
-            } catch (Exception exception) {
-                log.error("Failed to move from inbound file to unstructured data store for EOD id: {}", inboundObject.getId(), exception);
-            }
+                                                                                                                batchSize);
+        AtomicInteger count = new AtomicInteger();
+
+        System.out.println("TMP: here 1");
```
Comment (karen-hedges) [inline src/main/java/uk/gov/hmcts/darts/datamanagement/service/impl/InboundToUnstructuredProcessorImpl.java:70]: You might to go through and clear out your System.out statements

File: src/main/java/uk/gov/hmcts/darts/util/AsyncUtil.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+@Slf4j
+public final class AsyncUtil {
+
+    public static void shutdownAndAwaitTermination(ExecutorService executorService, int timeout, TimeUnit unit) {
+        executorService.shutdown();
+        try {
+            if (!executorService.awaitTermination(timeout, unit)) {
+                executorService.shutdownNow();
+                if (!executorService.awaitTermination(60, TimeUnit.SECONDS)) {
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/util/AsyncUtil.java:19]: The default timeout is one hour (defined on the processor class this is just if that time out is reached we will try to force the shutdown by skipping executions)

[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/util/AsyncUtil.java:22]: The default timeout is one hour (defined on the processor class this is just if that time out is reached we will try to force the shutdown by skipping executions)

File: src/main/java/uk/gov/hmcts/darts/datamanagement/service/impl/InboundToUnstructuredProcessorImpl.java (hunk 1 above)
Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/datamanagement/service/impl/InboundToUnstructuredProcessorImpl.java]: Updated

**Review by karen-hedges — APPROVED** [review]

**Comment by jackmaloney** [comment]
Other than above comment, LGTM

File: src/main/resources/db/migration/common/V1_389__inbound_to_unstructured_batch_size.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+update darts.automated_task
+set batch_size = 100
```
[the remark this answers is not in the document]
Comment (Ben-Edwards-cgi) [reply src/main/resources/db/migration/common/V1_389__inbound_to_unstructured_batch_size.sql]: updated

File: src/main/java/uk/gov/hmcts/darts/datamanagement/service/impl/InboundToUnstructuredProcessorImpl.java [hunk 4]
```diff
@@ -53,27 +48,49 @@ public class InboundToUnstructuredProcessorImpl implements InboundToUnstructured
...
+        List<Callable<Void>> tasks = inboundList.stream()
+            .map(inboundObject -> (Callable<Void>) () -> {
+                log.debug("Processing Inbound to Unstructured record {} of {} with EOD {}",
+                          count.getAndIncrement(), inboundList.size(), inboundObject);
+
+                processInboundToUnstructured(inboundObject);
+                return null;
+            }).toList();
+        try {
+            AsyncUtil.invokeAllAwaitTermination(tasks, threads, 1, TimeUnit.HOURS);
```
Comment (karen-hedges) [inline src/main/java/uk/gov/hmcts/darts/datamanagement/service/impl/InboundToUnstructuredProcessorImpl.java:77]: Actually just thought if the batch size gets set to less than thread size you don't want it to wait for more threads than in the batch. Also if the number of objects returned is less than thread size its the same

Comment (Ben-Edwards-cgi) [reply src/main/java/uk/gov/hmcts/darts/datamanagement/service/impl/InboundToUnstructuredProcessorImpl.java:77]: Threads are just the number of threads that get spun up that the tasks can execute on.
The await termination waits for all tasks to completed not the threads so even if we have fewer tasks then we do threads, we should not get stuck

**Review by davet1985 — APPROVED** [review]

**Pull request merged**